T47: Add backend logic for signing up a user

jessicahe004 opened the pull request:
Handles the axios post request from /signup for the registration of new users. 
- creates random salt and also hashes client password. 
- inserts email, first name, last name, hashed password, and salt into the database. 

*still need to handle general errors: unique email error & potential db being down & or website crashing 
*need to add redirection when successfully signed up to /login or /

jessicahe004 committed:
backendsignup

jessicahe004 committed:
Merge branch 'frontend' into backend

jessicahe004 committed:
Merge branch 'main' of https://github.com/wewrap/togetherwewrap into backend

jessicahe004 committed:
Merge branch 'frontend' into backend

chinkevin25 commented on backend/models/user.ts:
```diff
@@ -0,0 +1,10 @@
+    salt: string; 
+    birthDate: Date; 
+}
+
```
Did you make this yourself? If so do we need a new interface for this? I thought we would be able to use the Prisma Model for this?

I know Prisma has their own types but i'm not sure how the are generated / extracted: https://www.prisma.io/docs/concepts/components/prisma-schema/data-model#type-definitions

jessicahe004 replied:
i think i'm just going to delete this file

chinkevin25 commented on backend/routes/user.ts:
```diff
@@ -0,0 +1,28 @@
+
+const db = prisma; 
+
+app.post('/signup', 
+passport.authenticate('local', {failureRedirect: '/signup', failureMessage: true }),
+ function(req, res) {
+    console.log(req.body)
+ });
```
I believe you're going to end up moving the route here right, just checking

jessicahe004 replied:
i will end up moving this

jessicahe004 committed:
deleted:    backend/models/user.ts
modified:   backend/routes/signUpAuth.ts
deleted:    backend/routes/user.ts
modified:   backend/routes/signUpAuth.ts

jessicahe004 committed:
backend

jessicahe004 committed:
signup

justinpchen94 commented on backend/routes/signUpAuth.ts:
```diff
@@ -0,0 +1,47 @@
+                firstName: req.body.firstName,
+                lastName: req.body.lastName, 
+                password: hashedPassword.toString("base64"),
+                salt: salt //fix salt ??? how to best generate salt using crypto 
```
you already generated it correctly on line 21!

also a cool javascript shorthand is whenever you have an object with fields like this:

```
{
  salt: salt,
}
```

it can be rewritten as

```
{
  salt,
}
```

justinpchen94 replied:
see: https://javascript.plainenglish.io/object-literals-using-object-property-shorthand-6360825c60ef#:~:text=When%20ES6%20was%20created%20object,the%20name%20and%20the%20key.&text=myKeyName-,%7D%3B,us%20to%20write%20less%20code!

justinpchen94 commented on backend/routes/signUpAuth.ts:
```diff
@@ -0,0 +1,47 @@
+    res.send(200);
+});
+
+//error caused by emails should be unique but using same email fix error 
+// error states, styling forms, fix routes, deicide what needs to be sent back to the client (what's the next step) (reroute to a specific page), 
+
+// app.post('/signup', 
+// passport.authenticate('local', {failureRedirect: '/signup', failureMessage: true }),
+//  function(req, res) {
+//     console.log(req.body)
+//  });
```
let's rm if not needed!

justinpchen94 commented on backend/utils/app.ts:
```diff
@@ -0,0 +1,5 @@
+import express from 'express';
```
i think this file is not needed since we have a diff app.ts in the backend/ dir already I believe!

justinpchen94 commented on backend/utils/db.ts:
```diff
@@ -0,0 +1,6 @@
+import { PrismaClient } from '@prisma/client';
```
same, i think this logic is already in backend/app.ts

jessicahe004 replied:
okay ty i will fix that

jessicahe004 committed:
deleted files and comments

justinpchen94 commented on backend/routes/signUpAuth.ts:
```diff
@@ -0,0 +1,36 @@
+                salt 
+            },
+        })
+        console.log(user); //redirect to the login page 
```
let's add this as a TODO

justinpchen94 commented on backend/routes/signUpAuth.ts:
```diff
@@ -0,0 +1,36 @@
+        })
+        console.log(user); //redirect to the login page 
+    })
+    res.send(200);
```
what happens if the user.create fails for some reason?

jessicahe004 replied:
send an error code to frontend and then frontend will issue a warning that says issue with submission try again

justinpchen94 reviewed (approved):
small nits! nothing blocking on this pr
